Add plugin: Media Viewer

Devon22 opened the pull request:
# I am submitting a new Community Plugin

## Repo URL

<!--- Paste a link to your repo here for easy access -->
Link to my plugin: https://github.com/Devon22/obsidian-mediaviewer

## Release Checklist
- [ ] I have tested the plugin on
  - [x]  Windows
  - [ ]  macOS
  - [ ]  Linux
  - [x]  Android _(if applicable)_
  - [x]  iOS _(if applicable)_
- [ ] My GitHub release contains all required files (as individual files, not just in the source.zip / source.tar.gz)
  - [x] `main.js`
  - [x] `manifest.json`
  - [ ] `styles.css` _(optional)_
- [x] GitHub release name matches the exact version number specified in my manifest.json (_**Note:** Use the exact version number, don't include a prefix `v`_)
- [x] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x] I have read the developer policies at https://docs.obsidian.md/Developer+policies, and have assessed my plugins's adherence to these policies.
- [x] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x] I have added a license in the LICENSE file.
- [x] My project respects and is compatible with the original license of any code from other plugins that I'm using.
      I have given proper attribution to these other projects in my `README.md`.

joethei commented:
["description": "A media viewer plugin, allowing you to view and manage media files within your notes.",](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/manifest.json#L6)
Avoid including sentences like "This is a plugin that does ..." in your description, it should be a given since this is only shown when browsing plugins through the app or on our website.
We have published a guide for plugin descriptions to follow here: https://docs.obsidian.md/Plugins/Releasing/Submission+requirements+for+plugins#Keep+plugin+descriptions+short+and+simple.

['ALLOW_MEDIA_DELETION': 'Allow Media Deletion',](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L51), ['AUTO_OPEN_FIRST_IMAGE': 'Auto Open First Image',](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L53), ['CLICK_TO_OPEN_MEDIA': 'Click Image to Open Media Browser',](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L55), ['MUTE_VIDEO_ON_OPEN': 'Mute Video on Open',](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L57), ['GALLERY_GRID_WIDTH': 'Gallery Grid Width',](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L59), ['OPEN_MEDIA_VIEWER': 'Open Media Viewer',](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L63-L65)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

[videoIcon.innerHTML = '<svg viewBox="0 0 24 24" width="24" height="24"><path fill="currentColor" d="M8 5v14l11-7z"/></svg>';](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L296), [videoIcon.innerHTML = '<svg viewBox="0 0 24 24" width="24" height="24"><path fill="currentColor" d="M8 5v14l11-7z"/></svg>';](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L880), [noteIcon.innerHTML = `](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L1755-L1759), [noteIcon.innerHTML = `](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L1762-L1766), [videoIcon.innerHTML = '<svg viewBox="0 0 24 24" width="24" height="24"><path fill="currentColor" d="M8 5v14l11-7z"/></svg>';](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L880), [videoIcon.innerHTML = `](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L1824-L1828)
Don't use `innerHTML` for this, see the docs for what to do instead: https://docs.obsidian.md/Plugins/User+interface/Icons

[const style = document.createElement('style');](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L558-L798), [const styleEl = document.createElement('style');](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L1227-L1435)
All the hardcoded stuff in here should be in the `styles.css` file.

You have this system for supporting different languages, but a lot of strings are only in chinese, please enable translation for thiese.

Devon22 commented:
> ["description": "A media viewer plugin, allowing you to view and manage media files within your notes.",](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/manifest.json#L6) Avoid including sentences like "This is a plugin that does ..." in your description, it should be a given since this is only shown when browsing plugins through the app or on our website. We have published a guide for plugin descriptions to follow here: https://docs.obsidian.md/Plugins/Releasing/Submission+requirements+for+plugins#Keep+plugin+descriptions+short+and+simple.

Fixed.

> ['ALLOW_MEDIA_DELETION': 'Allow Media Deletion',](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L51), ['AUTO_OPEN_FIRST_IMAGE': 'Auto Open First Image',](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L53), ['CLICK_TO_OPEN_MEDIA': 'Click Image to Open Media Browser',](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L55), ['MUTE_VIDEO_ON_OPEN': 'Mute Video on Open',](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L57), ['GALLERY_GRID_WIDTH': 'Gallery Grid Width',](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L59), ['OPEN_MEDIA_VIEWER': 'Open Media Viewer',](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L63-L65) [Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

'GALLERY_GRID_WIDTH': 'Gallery Grid Width'
the latter part is what's displayed on the UI.

> [videoIcon.innerHTML = '';](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L296), [videoIcon.innerHTML = '';](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L880), [noteIcon.innerHTML = `](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L1755-L1759), [noteIcon.innerHTML = `](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L1762-L1766), [videoIcon.innerHTML = '';](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L880), [videoIcon.innerHTML = `](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L1824-L1828) Don't use `innerHTML` for this, see the docs for what to do instead: https://docs.obsidian.md/Plugins/User+interface/Icons
> 
> [const style = document.createElement('style');](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L558-L798), [const styleEl = document.createElement('style');](https://github.com/Devon22/obsidian-mediaviewer/blob/87300f26dd64aab533c8e486c4d3629f07b6ee23/main_src.js#L1227-L1435) All the hardcoded stuff in here should be in the `styles.css` file.

Fixed.

> You have this system for supporting different languages, but a lot of strings are only in chinese, please enable translation for thiese.

Fixed, only comments remain.

joethei commented:
[# Obsidian Media Viewer](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/README.md#L2)
Naming something "Obsidian xyz" is reserved for first party products we create, please change this header.

['ALLOW_MEDIA_DELETION': 'Allow Media Deletion',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L62), ['AUTO_OPEN_FIRST_IMAGE': 'Auto Open First Image',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L64), ['CLICK_TO_OPEN_MEDIA': 'Click Image to Open Media Browser',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L66), ['MUTE_VIDEO_ON_OPEN': 'Mute Video on Open',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L68), ['GALLERY_GRID_LARGE': 'Gallery Large Size',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L70), ['GALLERY_GRID_MEDIUM': 'Gallery Medium Size (Default)',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L72), ['GALLERY_GRID_SMALL': 'Gallery Small Size',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L74), ['OPEN_MEDIA_VIEWER': 'Open Media Viewer',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L78), ['GENERATE_GALLERY': 'Generate Gallery Block'](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L79), ['ADD_IMAGE': 'Add Image',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L59)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

[this.modalEl.style.background = openType === 'command' ? 'rgba(0, 0, 0, 0.5)' : 'rgba(0, 0, 0, 0.0)';](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L169)
This should not be hardcoded like this, it should be a CSS class that is applied, and use the Obsidian color variables.

[const internalMatches = Array.from(contentWithoutFrontmatter.matchAll(/!\[\[(.*?)(?:\|.*?)?\]\]/g));](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L196-L200), [const externalMatches = Array.from(contentWithoutFrontmatter.matchAll(/!\[(.*?)\]\((.*?)\)/g));](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L227-L231)
Use the MetadataCache API for this, and then you also won't need to remove the frontmatter before

[// 分離 frontmatter 和內容](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L623-L637), [// 讀取當前文件內容](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L1686-L1797)
Use the `process` function to modfy a file without causing conflicts with other plugins.
(Your `minAppVersion` needs to be increased to use this API)

[callback: () => this.generateGallery()](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L895)
This should be a [conditional command](https://docs.obsidian.md/Plugins/User+interface/Commands#Conditional+commands)

[id: 'mediaview-generategallery',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L893)
Remove the plugin ID here, it's not needed.

Devon22 commented:
> [# Obsidian Media Viewer](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/README.md#L2) Naming something "Obsidian xyz" is reserved for first party products we create, please change this header.

Fixed.

> ['ALLOW_MEDIA_DELETION': 'Allow Media Deletion',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L62), ['AUTO_OPEN_FIRST_IMAGE': 'Auto Open First Image',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L64), ['CLICK_TO_OPEN_MEDIA': 'Click Image to Open Media Browser',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L66), ['MUTE_VIDEO_ON_OPEN': 'Mute Video on Open',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L68), ['GALLERY_GRID_LARGE': 'Gallery Large Size',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L70), ['GALLERY_GRID_MEDIUM': 'Gallery Medium Size (Default)',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L72), ['GALLERY_GRID_SMALL': 'Gallery Small Size',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L74), ['OPEN_MEDIA_VIEWER': 'Open Media Viewer',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L78), ['GENERATE_GALLERY': 'Generate Gallery Block'](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L79), ['ADD_IMAGE': 'Add Image',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L59) [Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

Fixed.

> [this.modalEl.style.background = openType === 'command' ? 'rgba(0, 0, 0, 0.5)' : 'rgba(0, 0, 0, 0.0)';](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L169) This should not be hardcoded like this, it should be a CSS class that is applied, and use the Obsidian color variables.

Fixed.

> [const internalMatches = Array.from(contentWithoutFrontmatter.matchAll(/![[(._?)(?:|._?)?]]/g));](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L196-L200), [const externalMatches = Array.from(contentWithoutFrontmatter.matchAll(/![(._?)]((._?))/g));](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L227-L231) Use the MetadataCache API for this, and then you also won't need to remove the frontmatter before

Because the links in the Gallery Block need to be parsed, Metadatacache cannot be used.

> [// 分離 frontmatter 和內容](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L623-L637), [// 讀取當前文件內容](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L1686-L1797) Use the `process` function to modfy a file without causing conflicts with other plugins. (Your `minAppVersion` needs to be increased to use this API)

Fixed.

> [callback: () => this.generateGallery()](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L895) This should be a [conditional command](https://docs.obsidian.md/Plugins/User+interface/Commands#Conditional+commands)

Fixed.

> [id: 'mediaview-generategallery',](https://github.com/Devon22/obsidian-mediaviewer/blob/53607719a6acb44b60ccdc602663bb35ae2c62c6/main_src.js#L893) Remove the plugin ID here, it's not needed.

Fixed.

joethei commented:
> Because the links in the Gallery Block need to be parsed, Metadatacache cannot be used.
Your code here scans everything, not just the content in the codeblock.

Devon22 commented:
> > Because the links in the Gallery Block need to be parsed, Metadatacache cannot be used.
> Your code here scans everything, not just the content in the codeblock.

Metadatacache cannot get the image link in the code block, so I did this
